File: forms/who-counter-referral-form/back-end-with-loco/src/engine/counter_referral_validator.rs
//! WHO Counter-Referral Form completeness validator.
//!
//! Pure function: validates the WHO Counter-Referral Form data for
//! completeness. Each rule in `counter_referral_rules` is checked first
//! against `applies()` to honour the conditional logic of the form (e.g.
//! "patient/family informed Yes → explanation required"). When a rule
//! applies, it must also be `is_satisfied()` for the form to be complete;
//! otherwise the rule is recorded as a missing field.

use std::collections::BTreeMap;

use crate::engine::counter_referral_rules::{counter_referral_rules, section_label};
use crate::engine::types::{
    AssessmentData, FiredRule, SectionCompleteness, ValidationResult,
};

pub fn validate_counter_referral(data: &AssessmentData) -> ValidationResult {
    let mut section_map: BTreeMap<String, SectionCompleteness> = BTreeMap::new();
    let mut missing: Vec<FiredRule> = Vec::new();
    let mut total_required: u32 = 0;
    let mut total_satisfied: u32 = 0;

    // Preserve insertion order of first occurrence per section.
    let mut section_order: Vec<String> = Vec::new();

    for rule in counter_referral_rules() {
        if !(rule.applies)(data) {
            continue;
        }
        total_required += 1;
        let satisfied = (rule.is_satisfied)(data);

        let key = rule.section.to_string();
        if !section_map.contains_key(&key) {
            section_order.push(key.clone());
            section_map.insert(
                key.clone(),
                SectionCompleteness {
                    section: rule.section.to_string(),
                    section_label: section_label(rule.section).to_string(),
                    required: 0,
                    satisfied: 0,
                    missing: Vec::new(),
                },
            );
        }
        let bucket = section_map.get_mut(&key).unwrap();
        bucket.required += 1;

        if satisfied {
            total_satisfied += 1;
            bucket.satisfied += 1;
        } else {
            let fired = FiredRule {
                id: rule.id.to_string(),
                section: rule.section.to_string(),
                description: rule.description.to_string(),
            };
            missing.push(fired.clone());
            bucket.missing.push(fired);
        }
    }

    let sections: Vec<SectionCompleteness> = section_order
        .into_iter()
        .filter_map(|k| section_map.remove(&k))
        .collect();

    ValidationResult {
        complete: missing.is_empty(),
        total_required,
        total_satisfied,
        sections,
        missing,
    }
}

File: forms/who-counter-referral-form/front-end-with-html/js/counter-referral-validator.js
import { counterReferralRules as rules } from './counter-referral-rules.js';

// Pure function: validates the WHO Counter-Referral Form data for completeness.
//
// Each rule in `counterReferralRules` is checked first against `applies()` to
// honour the conditional logic of the form (e.g. "patient/family informed
// Yes -> explanation required"). When a rule applies, it must also be
// `isSatisfied()` for the form to be complete; otherwise the rule is recorded
// as a missing field.
//
// The validator returns an aggregate ValidationResult plus a per-section
// breakdown so the report can highlight exactly which questions are
// outstanding for the patient.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').SectionCompleteness} SectionCompleteness
 * @typedef {import('./types.js').SectionKey} SectionKey
 * @typedef {import('./types.js').ValidationResult} ValidationResult
 */

/**
 * @param {AssessmentData} data
 * @returns {ValidationResult}
 */
function validateCounterReferral(data) {
  

  /** @type {Map<SectionKey, SectionCompleteness>} */
  const sectionMap = new Map();
  /** @type {FiredRule[]} */
  const missing = [];
  let totalRequired = 0;
  let totalSatisfied = 0;

  for (const rule of rules) {
    let applies = false;
    try {
      applies = rule.applies(data);
    } catch (e) {
      console.warn('Counter-referral rule ' + rule.id + ' applies() failed:', e);
    }
    if (!applies) continue;

    totalRequired++;

    let satisfied = false;
    try {
      satisfied = rule.isSatisfied(data);
    } catch (e) {
      console.warn('Counter-referral rule ' + rule.id + ' isSatisfied() failed:', e);
    }

    let bucket = sectionMap.get(rule.section);
    if (!bucket) {
      bucket = {
        section: rule.section,
        required: 0,
        satisfied: 0,
        missing: []
      };
    }
    bucket.required++;

    if (satisfied) {
      totalSatisfied++;
      bucket.satisfied++;
    } else {
      const fired = {
        id: rule.id,
        section: rule.section,
        description: rule.description
      };
      missing.push(fired);
      bucket.missing.push(fired);
    }

    sectionMap.set(rule.section, bucket);
  }

  const sections = Array.from(sectionMap.values());

  return {
    complete: missing.length === 0,
    totalRequired: totalRequired,
    totalSatisfied: totalSatisfied,
    sections: sections,
    missing: missing
  };
}

export { validateCounterReferral };
